sapicore/pipeline/simulation.py

```diff
@@ -0,0 +1,172 @@
+
+from argparse import ArgumentParser
+from datetime import datetime
+
+import torch
+from torch import Tensor
+from alive_progress import alive_bar
+
+from sapicore.pipeline import Pipeline
+from sapicore.model import Model
+from sapicore.engine.network import Network
+
+from sapicore.utils.seed import fix_random_seed
+from sapicore.utils.signals import extend_input_current
+from sapicore.utils.tensorboard import TensorboardWriter, HDFData
+from sapicore.utils.constants import DT, TIME_FORMAT, SEED
+from sapicore.utils.io import ensure_dir, save_yaml, log_settings
+
+from sapicore.tests import ROOT
+
+# default hardware device for this pipeline.
+DEVICE = "cpu" if not torch.cuda.is_available() else "cuda:0"
+
+
+class Simulator(Pipeline):
+    """A prototypical simulation pipeline, bringing together model specification, disk I/O, and visualization.
+    Initialize a network and configure the simulation by providing a dictionary or a path to a YAML.
+
+    Parameters
+    ----------
+    configuration: dict or str
+        Configuration dictionary or path to a configuration YAML. If the path provided is relative, the simulation
+        pipeline will assume that the root is `sapicore/tests` to support the execution of YAML-configured tests.
+
+    """
+
+    def __init__(self, configuration: dict | str = None, **kwargs):
+        super().__init__(configuration=configuration, **kwargs)
+
+        # set random seed for entire simulation and apply default logger configuration.
+        fix_random_seed(SEED)
+        log_settings()
+
+    def run(self):
+        """Run the simulation based on the given configuration.
+
+        Raises
+        ------
+        KeyError
+            If full path to project `root` directory is not provided as a key in the configuration YAML or dictionary.
+
+        """
+        # get full path to project root directory from configuration file.
+        root = self.configuration.get("root")
+        if not root:
+            raise KeyError("Configuration missing project 'root' key.")
+
+        elif not os.path.isabs(root):
+            # if full path not provided, the relative path is appended to sapinet/tests.
+            root = os.path.join(ROOT, root)
+
+        # create timestamped run directory.
+        stamp = datetime.now().strftime(TIME_FORMAT)
+        run_dir = ensure_dir(os.path.join(root, stamp))
+
+        # save a copy of the configuration to this run's directory for reference.
+        save_yaml(self.configuration, os.path.join(run_dir, self.configuration.get("identifier", "run_cfg") + ".yaml"))
+
+        # load simulation settings from configuration dictionary.
+        steps = self.configuration.get("simulation", {}).get("steps")
+        disk = self.configuration.get("simulation", {}).get("disk")
+        tensorboard = self.configuration.get("simulation", {}).get("tensorboard")
+
+        # create data and tensorboard subdirectories.
+        tb_dir = ensure_dir(os.path.join(run_dir, "tb"))
```

Comment:
Not super important rn, but it would be nice if we accepted `Path` objects whereever we accept string paths. Although I know that tree-config would probably make that a little harder.

Reply:
I'll open an issue for that after we merge (good idea, but a bit of a hassle--can be deferred).